File: BinarySearch/medianTwoSortedArrays.py
// Pseudocode
//class Solution:
//    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
//        A, B = nums1, nums2
//        total = len(nums1) + len(nums2)
//        half = total // 2
//        
//        if len(B) < len(A):
//            A, B = B, A
//        
//        l, r = 0, len(A) - 1
//        while True:
//            i = (l + r) // 2 // A
//            j = half - i - 2 // B
//        
//            Aleft = A[i] if i >= 0 else float("-infinity")
//            Aright = A[i + 1] if (i + 1) < len(A) else float("infinity")
//            Bleft = B[j] if j >= 0 else float("-infinity")
//            Bright = B[j + 1] if (j + 1) < len(B) else float("infinity")
//        
//            // partition is correct
//            if Aleft <= Bright and Bleft <= Aright:
//                // odd
//                if total % 2:
//                    return min(Aright, Bright)
//                // even
//                return (max(Aleft, Bleft) + min(Aright, Bright)) / 2
//            elif Aleft > Bright:
//                r = i - 1
//            else:
//                l = i + 1

// In essence, we are using binary search in list a to find
// the partition that satisfies the constraints of either partition in either list
// being less than the value right in front of the other list's partition.
class Solution {
    public static double findMedianSortedArrays(int[] a, int[] b){
        int total_length = a.length + b.length;
        int half = total_length / 2;

        // We set a to be the shorter one because 
        if(a.length > b.length){
            int[] holder = b;
            b = a;
            a = holder;
        }

        int i, j, l = 0, r = a.length - 1;
        double Aleft, Aright, Bleft, Bright;
        while(true){
            i = (l + r) / 2;
            j = (half - 1) - (i + 1);

            Aleft = i >= 0 ? a[i] : Double.POSITIVE_INFINITY;
            Aright = (i + 1) < a.length ? a[i + 1] : -1 * Double.POSITIVE_INFINITY;
            Bleft = j >= 0 ? b[j] : Double.POSITIVE_INFINITY;
            Bright = (j + 1) < b.length ? b[j + 1] : -1 * Double.POSITIVE_INFINITY;

            // If the partition is correct
            if(Aleft <= Bright && Bleft <= Aright){
                // if odd
                if((total_length % 2) != 0){
                    return Math.min(Aright, Bright);
                }
                // if even
                return (Math.max(Aleft, Bleft) + Math.min(Aright, Bright)) / 2.0;
            }

            else if(Aleft > Bright){
                r = i - 1;
            }
            else{
                l = i + 1;
            }
        }
    }

    public static void main(String[] args){
        int[] array1 = {1, 2, 3, 4};
        int[] array2 = {1, 2, 3, 4, 5, 6, 7, 8};
        double answer = findMedianSortedArrays(array1, array2);
        System.out.println(answer);
    }
}

File: BinarySearch/medianTwoSortedArrays.java
// Important visual:
//  
//                 x:        1 2 4 7 9 | 11 14 20 21
//             
//                 y:        3 5 6 8 | 12 15 18 19
//  
//                                     |
//                                     | 
//                                     v 
//  
// Combined Array:  1 2 3 4 5 6 7 8 9 | 11 12 14 15 18 19 20 21
// Gist: The goal is the find the correct parition in the arrays that allows us to immediately find the median. The
// numbers just around the partitions will be the numbers of interest. Finding the partition is done with binary search
// (although iteration may also work). Then from there getting the median will be algebriac.

class Solution {
     public static double findMedianSortedArrays(int[] a, int[] b){
         //if a length is greater than switch them so that input1 is smaller than b.
        if (a.length > b.length) {
            return findMedianSortedArrays(b, a);
        }
        int x = a.length;
        int y = b.length;

        int low = 0;
        int high = x;
        while (low <= high) {
            int partitionX = (low + high)/2;
            int partitionY = (x + y + 1)/2 - partitionX;

            //if partitionX is 0 it means nothing is there on left side. Use -INF for maxLeftX
            //if partitionX is length of input then there is nothing on right side. Use +INF for minRightX
            int maxLeftX = (partitionX == 0) ? Integer.MIN_VALUE : a[partitionX - 1];
            int minRightX = (partitionX == x) ? Integer.MAX_VALUE : a[partitionX];

            int maxLeftY = (partitionY == 0) ? Integer.MIN_VALUE : b[partitionY - 1];
            int minRightY = (partitionY == y) ? Integer.MAX_VALUE : b[partitionY];

            if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
                //We have partitioned array at correct place
                // Now get max of left elements and min of right elements to get the median in case of even length combined array size
                // or get max of left for odd length combined array size.
                if ((x + y) % 2 == 0) {
                    return ((double)Math.max(maxLeftX, maxLeftY) + Math.min(minRightX, minRightY))/2;
                } else {
                    return (double)Math.max(maxLeftX, maxLeftY);
                }
            } else if (maxLeftX > minRightY) { //we are too far on right side for partitionX. Go on left side.
                high = partitionX - 1;
            } else { //we are too far on left side for partitionX. Go on right side.
                low = partitionX + 1;
            }
        }
        return 0.0;

     }
 }


